<a href="https://colab.research.google.com/github/LiTaokui/TensorFlow_Toturial/blob/main/Dropout_EX_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
df = pd.read_csv("/content/sample_data/sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
46   0.0308  0.0339  0.0202  0.0889  0.1570  0.1750  0.0920  0.1353  0.1593   
124  0.0388  0.0324  0.0688  0.0898  0.1267  0.1515  0.2134  0.2613  0.2832   
148  0.0712  0.0901  0.1276  0.1497  0.1284  0.1165  0.1285  0.1684  0.1830   
162  0.0217  0.0152  0.0346  0.0346  0.0484  0.0526  0.0773  0.0862  0.1451   
138  0.0731  0.1249  0.1665  0.1496  0.1443  0.2770  0.2555  0.1712  0.0466   

         9   ...      51      52      53      54      55      56      57  \
46   0.2795  ...  0.0167  0.0127  0.0138  0.0090  0.0051  0.0029  0.0122   
124  0.2718  ...  0.0255  0.0071  0.0263  0.0079  0.0111  0.0107  0.0068   
148  0.2127  ...  0.0154  0.0156  0.0054  0.0030  0.0048  0.0087  0.0101   
162  0.2110  ...  0.0123  0.0067  0.0011  0.0026  0.0049  0.0029  0.0022   
138  0.1114  ...  0.0444  0.0230  0.0290  0.0141  0.0161  0.0177  0.0194   

         58      59  60  
46   0.0056  0.0020   R  
124  0.0097  0.0067   M  
148  0.0095  0.0068   M  
162  0.0022  0.0032   M  
138  0.0207  0.0057   M  

[5 rows x 61 columns]

In [4]:
df.shape

(208, 61)

In [5]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [6]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [7]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [8]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [9]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
6    1
190  0
48   1
100  0
39   1

In [10]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [11]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [13]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

## Wihout Dropout ###

In [14]:
import tensorflow as tf
from tensorflow import keras

In [15]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6740 - accuracy: 0.5577
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6456 - accuracy: 0.6603
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6176 - accuracy: 0.6859
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5897 - accuracy: 0.7564
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5518 - accuracy: 0.7308
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5290 - accuracy: 0.7564
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4922 - accuracy: 0.8013
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4623 - accuracy: 0.7949
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4241 - accuracy: 0.8397
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4189 - accuracy: 0.8397
Epoch 11/

In [16]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 8ms/step - loss: 1.0244 - accuracy: 0.8077


[1.0243778228759766, 0.807692289352417]

Training Accuracy >>> Test Accuracy

In [17]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[5.0307172e-11 9.9315351e-01 9.9957317e-01 5.9019123e-07 9.9999714e-01
 9.9994588e-01 2.4472445e-01 9.9999917e-01 7.4647978e-08 9.9999958e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [18]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [19]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.89      0.83        27
           1       0.86      0.72      0.78        25

    accuracy                           0.81        52
   macro avg       0.82      0.80      0.81        52
weighted avg       0.81      0.81      0.81        52



### With Dropout ###

In [20]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7249 - accuracy: 0.5449
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7187 - accuracy: 0.5449
Epoch 3/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7180 - accuracy: 0.5128
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7143 - accuracy: 0.4808
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6961 - accuracy: 0.5577
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6999 - accuracy: 0.5641
Epoch 7/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6910 - accuracy: 0.5385
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6835 - accuracy: 0.5321
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7046 - accuracy: 0.5000
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6995 - accuracy: 0.5064
Epoch 11/

In [21]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 8ms/step - loss: 0.3957 - accuracy: 0.7885


[0.3957103490829468, 0.7884615659713745]

In [22]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[2.7023927e-06 4.2946327e-01 7.9849851e-01 8.3902180e-03 9.3084633e-01
 7.9935443e-01 4.8608515e-01 9.3647277e-01 5.8761239e-03 9.4615972e-01]
[0. 0. 1. 0. 1. 1. 0. 1. 0. 1.]


In [23]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.74      0.93      0.82        27
           1       0.89      0.64      0.74        25

    accuracy                           0.79        52
   macro avg       0.81      0.78      0.78        52
weighted avg       0.81      0.79      0.78        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.81 